---
author: "Céline Chevalier and Anaı̈s Baudot"
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
    rmdformats::material:
        use_bookdown: true
        thumbnails: false
        df_print: kable
        code_folding: hide
        number_sections: yes
    pdf_document:
        number_sections: yes
        toc: yes
        toc_depth: 3
        keep_tex: no
---

```{r, echo=FALSE}
if (combinedD){
    title_var <- paste0("|
    | Step 9: DEA on the ", ifelse(combine_meth=="merge", "merged", "integrated"), " data -- ", DATASET)
} else {
    title_var <- paste0("|
    | Step 6: Differential expression analysis -- ", DATASET)
}
```

---
title: `r title_var`
---

<!-- Javascript for zooming on figures (adapted from: https://stackoverflow.com/questions/40401680) -->

<!-- Jquery import conflicts with DT::datatable so needs to be commented here -->
<!-- <script src = "https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script> -->
<!-- 
<style>
.zoomDiv {
  display: none;  
  position: fixed;
  top: 50%;
  left: 50%;
  z-index: 50;
  transform: translate(-50%, -50%);
  background-color: #FFFFFF;
  box-shadow: 0px 0px 50px #888888;
  width: fit-content;
  max-width: 90%;
  max-height: 90%;
  overflow: auto;
}

.zoomImg {
  width: 150%;
}
</style>

<script type = "text/javascript">
  $(document).ready(function() {
    $("body").prepend("<div class = \"zoomDiv\"><img src = \"\" class = \"zoomImg\"></div>");
    // onClick for all img except the zoomed one and link ones (filter)
    // use "img.zoom" and out.extra = "class = \"zoom\"" in chunk to specify manually which chunk images can be zoomed
    $("img:not(.zoomImg)").filter(":not(a *)").click(function() {
      $(".zoomImg").attr("src", $(this).attr("src"));
      $(".zoomDiv").show();
    })
    // onClick function for hiding div
    $("img.zoomImg").click(function() {
      $(".zoomDiv").hide();
    })
  })
</script> -->


```{r setup, include = FALSE}
options(knitr.purl.inline = TRUE)
knitr::opts_chunk$set(
  # code evaluation
  eval = TRUE,

  # text output
  echo = TRUE,
  results = "hold",
  warning = FALSE,
  error = FALSE,
  message = FALSE,
  strip.white = TRUE,

  # code decoration
  tidy.opts = list(width.cutoff = 90),
  comment = "",
  attr.output = ".numberLines",

  # plots
  fig.path = paste0(PATH_OUT_FIG, "/06_annotDEAviz_scenario_", scenario, "_"),
  fig.show = "asis",         # tuned to "hold" in multiple plots chunk
  dev = c("png", "pdf"),
  out.width = "50%",
  fig.width = 12,
  fig.height = 12,
  fig.align = "center"   # should be tuned to default in multiple plots chunk
)
```

```{r load-libraries, include = FALSE}
library(Seurat)
library(ggplot2)
library(tidyr)
library(Matrix)
library(dplyr)
library(kableExtra)
library(stringr)
library(DT)
library(scales)
library(RColorBrewer)
library(grDevices)
library(DElegate)
library(data.table)

source("./plotting_functions.R", local = TRUE)
source("./data_management.R", local = TRUE)
source("./checkDirHierarchy.R", local = TRUE)
```

```{r load-colors, eval=atlas, include = FALSE}
checkPath(file.path(PATH_ATLAS, "colorsSheet.csv"))
checkPath(PATH_ATLAS_FILE)

# Create color vectors
colors.table <- read.csv(file.path(PATH_ATLAS, "colorsSheet.csv"), header=T, comment.char="", as.is=T)
colors.celltype <- setNames(colors.table$celltype_color[!is.na(colors.table$celltype)], colors.table$celltype[!is.na(colors.table$celltype)])
colors.stage <- setNames(colors.table$development_stage_color[!is.na(colors.table$development_stage)], colors.table$development_stage[!is.na(colors.table$development_stage)])
```

```{r dir-managment-1, include=FALSE, eval=combinedD}
# PATHS
PATH_CLUSTER_TABLE <- paste0(PATH_ROOT, "/08_combineData/clustering_", DATASET, "_scenario.", scenario, "_meth.", clust_meth, "_res.", clust_res, ".csv")
PATH_RDS_FILE <- paste0(PATH_RDS_OBJECTS, "/07_preprocessed_", DATASET, "_scenario.", scenario, ".rds")
PATH_OUT_ANALYSIS <- file.path(PATH_ROOT, "08_combineData")
if (!dir.exists(PATH_OUT_ANALYSIS)) { dir.create(PATH_OUT_ANALYSIS) }
feature_file <- file.path(PATH_ROOT, "00_rawData", paste0("combined_", DATASET, "_features.tsv.gz"))

# check file existence
checkPath(PATH_CLUSTER_TABLE)
checkPath(PATH_RDS_FILE)
checkPath(feature_file)

if (!is.null(genes_of_interest)) {
    checkPath(genes_of_interest)
}
```

```{r dir-managment-2, include=FALSE, eval=!combinedD}
# PATHS
PATH_CLUSTER_TABLE <- paste0(PATH_ROOT, "/06_process/clustering_", DATASET, "_scenario.", scenario, "_meth.", clust_meth, "_res.", clust_res, ".csv")
PATH_RDS_FILE <- paste0(PATH_RDS_OBJECTS, "/05_preprocessed_", DATASET, "_scenario.", scenario, ".rds")
PATH_OUT_ANALYSIS <- file.path(PATH_ROOT, "07_annotDEAviz")
if (!dir.exists(PATH_OUT_ANALYSIS)) { dir.create(PATH_OUT_ANALYSIS) }
feature_file <- file.path(PATH_INPUT_LABDATA, "features.tsv.gz")

# check file existence
checkPath(PATH_CLUSTER_TABLE)
checkPath(PATH_RDS_FILE)
checkPath(feature_file)

if (!is.null(genes_of_interest)) {
    checkPath(genes_of_interest)
}
```

```{r load-data, include=FALSE}
# LOAD DATA
fltd <- readRDS(PATH_RDS_FILE)

# LOAD CLUSTERING RESULTS
clusters <- read.table(PATH_CLUSTER_TABLE, header = TRUE, sep = ",", row.names = "cellIDs")
clustOfInterest <- grep(paste0(DefaultAssay(fltd), "_snn_res.", clust_res), names(clusters), value = TRUE)
fltd@meta.data <- merge(fltd@meta.data, clusters[clustOfInterest], by = 'row.names')
fltd@meta.data <- tibble::column_to_rownames(fltd@meta.data, "Row.names")
```

```{r, eval=atlas, echo=FALSE, results="asis"}
cat("# Cells annotation through label transfer from the atlas\n")
```

```{r mapping-atlas, eval=atlas, child="60_cellTypeAnnotationTransfer.Rmd"}
```

<!-- ## Impact of the additional controls on the annotation transfer -->

<!-- ```{r heatmap-afterDF} -->
<!-- Idents(fltd) <- fltd@meta.data$celltype_DF -->
<!-- transition <- table(fltd@meta.data$celltype_DF, fltd@meta.data$celltype_singleData) -->
<!-- dt <- as.data.frame(transition, row.names = names(fltd@meta.data$celltype_singleData)) -->
<!-- dt[dt == 0] <- NA -->
<!-- names(dt) <- c("beforeDF", "afterDF", "counts") -->


<!-- ggplot(dt, aes(beforeDF, afterDF, fill = counts)) + # nolint -->
<!--     geom_tile() + -->
<!--     geom_text(aes(label = counts), color = "black", size = 2.5, fontface = "bold") + -->
<!--     #scale_color_manual(values = c(NA, "black"), na.value = NA) + -->
<!--     coord_fixed() + -->
<!--     scale_fill_gradient(low = "white", high = "pink", na.value="white") + -->
<!--     theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) + -->
<!--     guides(fill = guide_colourbar(barwidth = 0.5, -->
<!--                                   barheight = 10, -->
<!--                                   title = "Counts")) + -->
<!--     ggtitle(paste0(DATASET, " : impact of doublets removal on the label transfer")) + -->
<!--     CenterTitle() + -->
<!--     labs(x = "Cell type annotation before additional controls", y = paste0("Cell type annotation after additional controls")) -->
<!-- ``` -->

# Perform differential expression analysis

```{r run-dea}
# Increase the maximum amount of data able to be passed to future workers to 2GB
options(future.globals.maxSize = 2 * 1024^3)

corr_table <- as.data.frame(fread(feature_file, header = FALSE, sep = "\t", col.names = c("ENSid", "GeneName", "Type")))

clustOfInterest <- grep(paste0(DefaultAssay(fltd), "_snn_res.", clust_res), names(fltd@meta.data), value = TRUE)
Idents(fltd) <- clustOfInterest

DefaultAssay(fltd) <- 'RNA'

# Run DEA analysis
set.seed(general_seed)
# These parameters should be moved to the config file eventually
fc_threshold <- 0
padj_threshold <- 0.001
# Performs DEA analysis with the "each vs rest" method :
# For each cluster, the cells are randomly split in 3 groups of pseudoreplicates then aggregated as pseudobulk
# Then the rest of the cells are pooled together, split and pseudobulked in the same way, before running DESeq2 to find the DE genes.
markers <- FindAllMarkers2(fltd, method="deseq", min_fc = fc_threshold)

# filter in significant marker genes and order them
markers.sign <- markers[markers$padj < padj_threshold, ]
markers.sign <- markers.sign[order(-markers.sign$log_fc), ]
markers.sign$cluster <- factor(markers.sign$group1, levels = as.character(sort(as.numeric(levels(as.factor(markers.sign$group1))))))
markers.sign <- markers.sign[order(markers.sign$cluster), ]

# Add gene names and cluster column
markers.sign$gene_name <- corr_table$GeneName[match(markers.sign$feature, corr_table$ENSid)]

# Order the output table
markers.sign <- markers.sign[ , c(12, 11, 3, 7, 8, 5, 6, 1)]

# Export the table
write.table(x = markers.sign, file = file.path(PATH_OUT_ANALYSIS,
                                               paste0("DESeq2_markers_min_fc_", fc_threshold, "_pad_", padj_threshold, "_", 
                                                      DATASET, "_scenario.", scenario, "_res.", clust_res, ".csv")), sep = ",", row.names = F, col.names = T)

# Display top markers
markers.sign$cluster <- as.numeric(markers.sign$cluster)
topNmarkers <- extract_top_features(markers.sign, topn = top_markers)
datatable(topNmarkers, rownames = FALSE, filter = "top", options = list(pageLength = 5, scrollX=T))

# write.table(x = topNmarkers, file = file.path(PATH_OUT_ANALYSIS,
#                                               paste0("top_", top_markers, "_markers_", DATASET, "_scenario.", scenario, "_res.", clust_res, ".csv")), sep = ",", row.names = F, col.names = T)
write.table(x = topNmarkers, file = file.path(PATH_OUT_ANALYSIS,
                                              paste0("DESeq2_top_", top_markers, "_markers_", DATASET, "_scenario.", scenario, "_res.", clust_res, ".csv")), sep = ",", row.names = F, col.names = T)
```

# Additional Plots

```{r load-goi-list, eval=!is.null(genes_of_interest), echo=!is.null(genes_of_interest)}
# Load list of genes of interest
goi <- readLines(genes_of_interest)
goiList <- data.frame(GeneName = goi, ENSid = corr_table$ENSid[match(goi, corr_table$GeneName)])

# get gene_name that do not match ensembl id
genesToRemove <- goiList$GeneName[is.na(goiList$ENSid)]

if (length(genesToRemove) == dim(goiList)[1]){
	cat("None of the genes of interest are in the dataset")
	rm(goiList)
} else if (length(genesToRemove) >= 1){
    cat(paste0("One or more genes of interest are not in the dataset (maybe because of the case):\n\t",
               paste(genesToRemove, collapse = ", "), "\nThe listed genes will not be plotted.
               "))
	goiList <- na.omit(goiList) # remove the unknown genes
}
```

## Cluster identity {.tabset .tabset-pills .tabset-fade}

```{r get-topNmarkers}
# Parameters
ntop <- 5

topNmarkers <- extract_top_features(markers.sign, topn = ntop)
topNgenes <- unique(topNmarkers$feature)
Idents(fltd) <- factor(fltd@meta.data[,clustOfInterest], levels = sort(as.numeric(levels(as.factor(fltd@meta.data[,clustOfInterest])))))
```

### DimPlot of the cluster identity
```{r cluster-DimPlot-markers, out.width="50%"}
p <- DimPlot(fltd, reduction = "umap", pt.size = 0.8,
             label = FALSE) +
  ggtitle(paste0("Clusters at resolution ", clust_res, "\n", DATASET)) +
  CenterTitle() +
  NoAxes()
LabelClusters(p, id = "ident",  fontface = "bold", size = 5)
```

### DotPlot of the markers on cluster identity

```{r cluster-DotPlot-markers}
topNgenes <- data.frame(ENSid = topNgenes, GeneName = corr_table$GeneName[match(topNgenes, corr_table$ENSid)])
create_dot_plot(fltd, topNgenes, title=paste0("Top ", ntop, " markers of each cluster in clusters"))
```

```{r cluster-DotPlot-goi, eval=exists("goiList"), echo=exists("goiList"), results='asis'}
cat("### DotPlot of the genes of interest on cluster identity
")

create_dot_plot(fltd, goiList, title = "Genes of interest in clusters")
```

### FeaturePlots of the markers {.tabset .tabset-pills .tabset-fade}

```{r cluster-FeaturePlots-markers, results='asis', out.width='20%', fig.align='default'}
# code with the courtesy of Lionel Spinelli and Romain Fenouil
topMarkers = by( markers.sign, markers.sign[["cluster"]], function(x)
{
  # Filter markers based on adjusted PValue
  x = x[ x[["padj"]] < 0.005, , drop = FALSE]
  # Sort by decreasing logFC
  x = x[ order(abs(x[["log_fc"]]), decreasing = TRUE), , drop = FALSE ]
  rownames(x) <- x[["feature"]]
  # Return top ones
  return( if(is.null(  top_markers )) head( x, n =  10 ) else head( x, n =  min(top_markers, 10) ))
})


clustersColor <- setNames(hue_pal()(length(names(topMarkers))), names(topMarkers))

invisible( lapply( names( topMarkers), function(clusterName)
{
  cat("#### Cl. <span style='border-radius: 3px border: 3px solid ", clustersColor[clusterName], " padding:0px 2px'>", clusterName, "</span>\n")
  
  # Highlight cells of current cluster on a dimreduc plot
  highlightClusterPlot(clusterName, seuratObject = fltd, reduction = ifelse( exists("useReduction"), useReduction, "umap"))
  
  # Plots expression on projected cells
  invisible( lapply( topMarkers[[clusterName]][["feature"]], function(featureName)
  {
    print( FeaturePlot( fltd, features = featureName, reduction = ifelse( exists("useReduction"), useReduction, "umap"), order = TRUE) +
            theme( axis.title.x = element_blank(),
                    axis.title.y = element_blank(),
                    legend.position = "none") +
    labs(title = topMarkers[[clusterName]][[featureName, "gene_name"]]))
  }))
  
  cat(" \n \n") # Required for '.tabset'
}))
```

```{r cluster-FeaturePlot-goi, out.width='20%', fig.align='default', eval=exists("goiList"), echo=exists("goiList"), results='asis'}
cat("### FeaturePlots of the genes of interest with cluster labels
")

makeFeaturePlot <- function(geneOI){
	print(FeaturePlot(fltd,
    features = geneOI[["ENSid"]],
    reduction = ifelse(exists("useReduction"), useReduction, "umap"),
    min.cutoff = "q01", max.cutoff = "q99",
    order = TRUE, label = TRUE) +
    labs(title = geneOI[["GeneName"]]))

}

# Check if the genes of interrest are indeed in the filtered data
available_genes <- rownames(fltd@assays$RNA@data)
undetected_genes <- goiList[!goiList$ENSid %in% available_genes, "GeneName"]

# Print the undetected genes in the report 
if(length(undetected_genes) > 0) {
    cat("#### Undetected Genes of Interest\n")
    cat("The following genes present in the provided file were not detected in the dataset:\n")
    cat(paste(undetected_genes, collapse = ", "), "\n\n")
}

# Filter the genes of interrest list to the detected ones
goiList <- goiList[goiList$ENSid %in% available_genes, ]

# FeaturePlot the genes of interest
invisible(apply(goiList, 1, makeFeaturePlot))
```

### ViolinPlots of the markers {.tabset .tabset-pills .tabset-fade}

```{r clusters-VlnPlots-markers, results='asis', out.width='20%', fig.align='default'}
# with courtesy of Lionel Spinelli and Romain Fenouil
invisible( lapply( names( topMarkers), function(clusterName)
{
  cat("#### Cl. <span style='border-radius: 3px border: 3px solid ", clustersColor[clusterName], " padding:0px 2px'>", clusterName, "</span>\n")
  
  # Remind cluster name in an empty figure to keep consistent alignment of panels between tabs
  plot( c( 0, 1), c( 0, 1), ann = F, bty = 'n', type = 'n', xaxt = 'n', yaxt = 'n')
  text( x = 0.5, y = 0.5, paste( "Cluster", clusterName), cex = 2, col = clustersColor[clusterName])
  
  # Violinplot for expression value of marker genes by cluster (+ number of 'zero' and 'not zero' cells)
  invisible( lapply( topMarkers[[clusterName]][["feature"]], violinFeatureByCluster, seuratObject = fltd, clustersColor = clustersColor))
  
  cat(" \n \n") # Required for '.tabset'
}))
```

```{r set-celltype-identity, eval=atlas, results='asis'}
cat("## Celltype identity {.tabset .tabset-pills .tabset-fade}")
Idents(fltd) <- factor(fltd@meta.data[, colname], levels = sort(levels(as.factor(fltd@meta.data[, colname]))))
```

```{r celltype-DimPlot, eval=atlas, results='asis', out.width="50%", results='asis'}
cat("### DimPlot of the celltype identity {.tabset .tabset-pills .tabset-fade}")
## DimPlot the celltype
DimPlot(fltd,
		reduction = "umap",
        pt.size = 1,
        repel = TRUE,
        label = TRUE,
        label.size = 6,
        cols = colors.celltype[levels(Idents(fltd))]) +
		ggtitle(paste0(DATASET, " : cell identities")) +
		defaultTheme() +
		NoLegend()

cellsData <- data.frame(fltd@reductions[["umap"]]@cell.embeddings, fltd@meta.data[, colname])
colnames(cellsData) <- c(colnames(fltd@reductions[["umap"]]@cell.embeddings), "celltype")

invisible(sapply(levels(as.factor(fltd@meta.data[, colname])), function(act)
{
    cat("#### ", act, "\n")
        
    print(ggplot(cellsData[c('UMAP_1', 'UMAP_2')], # Omit the column used for facetting to get all points repeated in all facets
         aes( x = UMAP_1,
              y = UMAP_2)) +
    geom_point( alpha = .4,
                size  = 1,
                color = "grey") +
    geom_point( data = cellsData[cellsData$celltype == act,], # Now provide data including column for facetting
                aes(color = celltype),
                alpha = .5,
                size  = 3.5) +
    blankTheme() +
    scale_fill_manual(values = colors.celltype[fltd@meta.data[, colname]], aesthetics = "color") +
    NoLegend() +
    ggtitle(paste0(act))
    )

    cat(" \n \n") # Required for '.tabset'
}))
```

```{r celltype-DotPlot-markers, eval=atlas, results='asis'}
cat("### DotPlot of the markers on celltype identity")
## DotPlot the n-top markers of each celltype
create_dot_plot(fltd, topNgenes, title=paste0("Top ", ntop, " markers of each cluster in celltypes"))
```

```{r celltype-DotPlot-goi, eval=atlas, results='asis'}
cat("### DotPlot of the genes of interest on celltype identity")
if(!is.null(genes_of_interest)){
## DotPlot the genes of interest
create_dot_plot(fltd, goiList, title = "Genes of interest in celltypes")
} else {
  cat("These are not the droids you are looking for.")
}
```


<!-- ## Personal exploration -->

<!-- List of genes obtained from the CHD-RF-KB (A knowledgebase for non-syndromic -->
<!-- congenital heart disease associated risk factors). -->

<!-- ```{r} -->
<!-- # Load list of genes of interest -->
<!-- goi <- read.table(file.path(PATH_REQUIREMENTS, "geneList_CHD-RF-KB.txt")) -->
<!-- goiList <- goi$V1 -->
<!-- goiList <- str_to_title(goiList) -->

<!-- genesToRemove <- goiList[!(goiList %in% rownames(fltd))] -->
<!-- if (length(genesToRemove) >= 1){ -->
<!--   cat(paste0("One or more genes of interest are not in the dataset (maybe because of the case):\n\t", -->
<!--              paste(genesToRemove, collapse = ", "), "\nThe listed genes will not be plotted")) -->
<!-- } -->

<!-- goiList <- goiList[goiList %in% rownames(fltd)] # remove the unknown genes -->

<!-- Idents(fltd) <- factor(fltd@meta.data[,clustOfInterest], levels = sort(as.numeric(levels(as.factor(fltd@meta.data[,clustOfInterest]))))) -->
<!-- dotPlotsGoi(fltd, goiList, title="Genes of CHD-RF-KB database") -->
<!-- ``` -->


# SessionInfo

```{r sessionInfo}
sessionInfo()
```
